packages/graphql/src/schema-model/generate-model.ts

```diff
@@ -268,52 +279,109 @@ function hydrateRelationshipDeclarations(
+}
+
+/**
+ * Goes up the inheritance chain checking for the field to have the @relationshipDeclaration directive
+ * Finds the first interface that declares the field as a relationship
+ * Returns the name of the first interface and the target of the relationship declaration in that first interface
+ *
+ * @param definition Entity with relationship field (Starting point)
+ * @param fieldName Relationship field name (The one we look for first declaration for)
+ * @param definitionCollection
+ * @param schema
+ * @returns Info about the interface at the top of the chain, nullable because there might not be any
+ */
+function getFirstDeclaration(
     definition: ObjectTypeDefinitionNode | InterfaceTypeDefinitionNode,
     fieldName: string,
-    definitionCollection: DefinitionCollection
-): string | undefined {
-    // TODO: potentially use this instead
-    // const fieldNameToSourceNameMap = definition.interfaces?.reduce((acc, interfaceNamedNode) => {
-    //     const interfaceName = interfaceNamedNode.name.value;
-    //     const fields = definitionCollection.interfaceTypes.get(interfaceName)?.fields || [];
-    //     fields.forEach((f) => {
-    //         const exists = acc.has(f.name.value);
-    //         if (!exists) {
-    //             acc.set(f.name.value, interfaceName);
-    //         }
-    //     });
-    //     return acc;
-    // }, new Map<string, string>());
-
-    // deliberately using the first interface ONLY
-    const fieldNameToSourceNameMap = new Map<string, string>();
-    const firstInterfaceName = definition.interfaces?.[0]?.name.value;
-    if (firstInterfaceName) {
-        const fields = definitionCollection.interfaceTypes.get(firstInterfaceName)?.fields || [];
-        fields.forEach((field) => fieldNameToSourceNameMap.set(field.name.value, firstInterfaceName));
+    definitionCollection: DefinitionCollection,
+    schema: Neo4jGraphQLSchemaModel
+): { originalTarget?: Entity; firstDeclaredInTypeName?: string } {
+    if (!definition.interfaces) {
+        return {};
     }
-    return fieldNameToSourceNameMap?.get(fieldName);
+    const { inheritedInterfaceWithDeclaredField, declaredFieldTypeName } = definition.interfaces.reduce(
+        (acc, interfaceNamedNode) => {
+            const interfaceDef = getDefinitionNodeFromNamedNode(interfaceNamedNode, definitionCollection);
+            const declaredRelationshipField = getFieldDeclaredAsRelationship(interfaceDef, fieldName);
+            if (declaredRelationshipField) {
+                return {
+                    inheritedInterfaceWithDeclaredField: interfaceNamedNode,
+                    declaredFieldTypeName: getInnerTypeName(declaredRelationshipField.type),
+                };
+            }
+            return acc;
+        },
+        {} as {
+            inheritedInterfaceWithDeclaredField: NamedTypeNode | undefined;
+            declaredFieldTypeName: string | undefined;
+        }
+    );
```

Comment:
```suggestion
    let inheritedInterfaceWithDeclaredField: NamedTypeNode | undefined = undefined;
    let declaredFieldTypeName: string | undefined = undefined;

    for (const interfaceNamedNode of definition.interfaces) {
        const interfaceDef = getDefinitionNodeFromNamedNode(interfaceNamedNode, definitionCollection);
        const declaredRelationshipField = getFieldDeclaredAsRelationship(interfaceDef, fieldName);
        if (declaredRelationshipField) {
            inheritedInterfaceWithDeclaredField = interfaceNamedNode;
            declaredFieldTypeName = getInnerTypeName(declaredRelationshipField.type);
        }
    }
```

I think the cognitive complexity and type safety can be improved a little by changing this reduce to a for-of loop